src/main/java/com/cgm/infolab/db/ChatMessageRepository.java

```diff
@@ -0,0 +1,87 @@
+package com.cgm.infolab.db;
+
+import com.cgm.infolab.model.ChatMessage;
+import com.cgm.infolab.model.Room;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.dao.DuplicateKeyException;
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.simple.SimpleJdbcInsert;
+import org.springframework.stereotype.Component;
+
+import javax.sql.DataSource;
+import java.sql.Timestamp;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@Component
+public class ChatMessageRepository {
+    private final JdbcTemplate jdbcTemplate;
+    private final DataSource dataSource;
+    private final UserRepository userRepository;
+    private final RoomRepository roomRepository;
+
+    private final Logger log = LoggerFactory.getLogger(ChatMessageRepository.class);
+
+    public ChatMessageRepository(JdbcTemplate jdbcTemplate, DataSource dataSource,
+                                 UserRepository userRepository, RoomRepository roomRepository) {
+        this.jdbcTemplate = jdbcTemplate;
+        this.dataSource = dataSource;
+        this.userRepository = userRepository;
+        this.roomRepository = roomRepository;
+    }
+
+    /**
+     * Metodo che aggiunge un messaggio al database.
+     * @param message messaggio da salvare.
+     * @return chiave che è stata auto generata per il messaggio creato, oppure -1 se il messaggio inserito esisteva già.
+     */
+    public long add(ChatMessage message) {
+        SimpleJdbcInsert simpleJdbcInsert = new SimpleJdbcInsert(dataSource)
+                .withSchemaName("infolab")
+                .withTableName("chatmessages")
+                .usingGeneratedKeyColumns("id");
+
+        Map<String, Object> parameters = new HashMap<>();
+        parameters.put("sender_id", userRepository.readId(message.getUserSender()));
+        // TODO: trovare un modo per prendere la room in cui il messaggio arriva.
+        parameters.put("recipient_room_id", roomRepository.readId(new Room("general")));
+        parameters.put("sent_at", new Timestamp(System.currentTimeMillis()));
+        parameters.put("content", message.getContent());
+
+        try {
+            return (long)simpleJdbcInsert.executeAndReturnKey(parameters);
+        } catch (DuplicateKeyException e) {
```

Comment:
Anche qui gestire l'eccezione internamente al repository non è il massimo dal punto di vista di riusabilità di questo codice. Ne limita l'utilizzo nei contesti in cui vi va bene sopprimere la notifica di un duplicato, e non esserne informato per gestirlo in maniera diversa. Lascerei andare l'eccezione e la gestirei nel controller